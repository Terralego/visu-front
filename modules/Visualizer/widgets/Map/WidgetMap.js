import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import mapBoxGl from 'mapbox-gl';
import debounce from 'lodash.debounce';

import { toggleLayerVisibility, addListenerOnLayer, setLayerOpacity } from '../../services/mapUtils';
import LayersTreeProps from '../../propTypes/LayersTreePropTypes';
import DefaultMapComponent from './components/Map';
import LayersTree from './components/LayersTree';
import MarkdownRenderer from '../../components/MarkdownRenderer';
import Legend from './components/Legend';

import './styles.scss';

export const INTERACTION_DISPLAY_DETAILS = 'displayDetails';
export const INTERACTION_DISPLAY_TOOLTIP = 'displayTooltip';
export const INTERACTION_FN = 'function';

export class WidgetMap extends React.Component {
  static propTypes = {
    layersTree: LayersTreeProps,
    interactions: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      trigger: PropTypes.oneOf(['click', 'mouseover']),
      interaction: PropTypes.oneOf([
        INTERACTION_DISPLAY_DETAILS,
        INTERACTION_DISPLAY_TOOLTIP,
        INTERACTION_FN,
      ]),
      // for INTERACTION_DISPLAY_DETAILS
      // and INTERACTION_DISPLAY_TOOLTIP
      template: PropTypes.string,
      content: PropTypes.string,
      // for INTERACTION_FN
      fn: PropTypes.func,
    })),
    LayersTreeComponent: PropTypes.func,
    MapComponent: PropTypes.func,
    setDetails: PropTypes.func,
  };

  static defaultProps = {
    layersTree: [],
    LayersTreeComponent: LayersTree,
    MapComponent: DefaultMapComponent,
    interactions: [],
    setDetails () {},
  };

  state = {
    legends: [],
  };

  mapRef = React.createRef();

  map = new Promise(resolve => {
    const waitForMap = () => {
      if (this.mapRef && this.mapRef.current && this.mapRef.current.map) {
        resolve(this.mapRef.current.map);
        return;
      }
      setTimeout(waitForMap, 10);
    };
    waitForMap();
  });

  popups = new Map();

  hideTooltip = debounce(({ layerId }) => {
    if (!this.popups.has(layerId)) {
      return;
    }
    const popup = this.popups.get(layerId);
    popup.remove();
    this.popups.delete(layerId);
  }, 100);

  componentDidMount () {
    this.setInteractions();
  }

  componentDidUpdate ({ interactions: prevInteractions }) {
    const { interactions } = this.props;

    if (interactions !== prevInteractions) {
      this.setInteractions();
    }
  }

  onChange = async ({ layer, state: { active, opacity } }) => {
    const map = await this.map;

    if (active !== undefined) {
      layer.layers.forEach(layerId => {
        toggleLayerVisibility(map, layerId, active ? 'visible' : 'none');
        this.toggleLegend(layer, active);
      });
    }
    if (opacity !== undefined) {
      layer.layers.forEach(layerId =>
        setLayerOpacity(map, layerId, opacity));
    }
  }

  async setInteractions () {
    const map = await this.map;
    const { interactions } = this.props;

    if (this.mapInteractionsListeners) {
      this.mapInteractionsListeners.forEach(off => off());
    }

    this.mapInteractionsListeners = [];

    interactions.forEach(({ id, trigger = 'click', interaction, fn, ...config }) => {
      const calls = [];
      switch (interaction) {
        case INTERACTION_DISPLAY_DETAILS:
          calls.push({
            callback: (layerId, features) => this.displayDetails({ features, ...config }),
            trigger,
            displayCursor: trigger === 'click',
          });
          break;
        case INTERACTION_DISPLAY_TOOLTIP:
          if (trigger === 'mouseover') {
            calls.push({
              callback: (layerId, features, event) =>
                this.displayTooltip({ layerId, features, event, ...config }),
              trigger: 'mousemove',
            });
            calls.push({
              callback: (layerId, features, event) =>
                this.hideTooltip({ layerId, features, event, ...config }),
              trigger: 'mouseleave',
            });
          } else {
            calls.push({
              callback: (layerId, features, event) =>
                this.displayTooltip({ layerId, features, event, ...config }),
              trigger,
              displayCursor: trigger === 'click',
            });
          }
          break;
        case INTERACTION_FN:
          calls.push({
            callback: fn,
            trigger,
            displayCursor: trigger === 'click',
          });
          break;
        default:
          return;
      }
      const listeners = calls.reduce((list, { callback, trigger: callTrigger, displayCursor }) => [
        ...list,
        ...addListenerOnLayer(
          map,
          id,
          callback,
          { displayCursor, trigger: callTrigger },
        ),
      ], []);

      this.mapInteractionsListeners = this.mapInteractionsListeners.concat(listeners);
    });
  }

  displayDetails = ({ features, template }) => {
    const { setDetails } = this.props;
    setDetails({ features, template });
  }

  displayTooltip = async ({
    layerId,
    features: [{ properties }] = [{}],
    event: { lngLat },
    template,
    content,
  }) => {
    const map = await this.map;
    const container = document.createElement('div');
    ReactDOM.render(
      <MarkdownRenderer
        template={template}
        content={content}
        {...properties}
      />,
      container,
    );

    if (this.popups.has(layerId)) {
      this.popups.get(layerId).setLngLat([lngLat.lng, lngLat.lat]);
      return;
    }
    const popup = new mapBoxGl.Popup();
    this.popups.set(layerId, popup);
    popup.setLngLat([lngLat.lng, lngLat.lat]);
    popup.setDOMContent(container);
    popup.addTo(map);
  }

  toggleLegend (layer, active) {
    const { legend } = layer;
    if (!legend) return;
    const { legends } = this.state;
    const pos = legends.findIndex(l => l === legend);
    if (active && pos === -1) {
      this.setState({
        legends: [...legends, legend],
      });
    }
    if (!active && pos > -1) {
      legends.splice(pos, 1);
      this.setState({
        legends: [...legends],
      });
    }
  }

  render () {
    const {
      LayersTreeComponent, MapComponent, layersTree, style, interactions, ...mapProps
    } = this.props;
    const { legends } = this.state;
    const { onChange, mapRef } = this;

    return (
      <div
        className="widget-map"
        style={style}
      >
        {!!layersTree.length && (
        <LayersTreeComponent
          layersTree={layersTree}
          onChange={onChange}
        />
        )}
        <MapComponent
          {...mapProps}
          ref={mapRef}
          onDetailsChange={this.onDetailsChange}
        />
        {!!legends.length && (
          <div className="widget-map__legends">
            {legends.map(legend => (
              <Legend {...legend} />
            ))}
          </div>
        )}
      </div>
    );
  }
}

export default WidgetMap;
