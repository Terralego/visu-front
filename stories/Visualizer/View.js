
import React from 'react';
import { storiesOf } from '@storybook/react';

import VisualizerProvider, { View } from '../../modules/Visualizer/';
import layersTree from './data/layersTree';

const stories = storiesOf('Module Visualizer');

stories.add('View Component', () => (
  <VisualizerProvider>
    <div style={{ height: '100vh' }}>
      <View
        widgets={[{
          type: 'map',
          layersTree,
          accessToken: 'pk.eyJ1IjoiaGFkcmllbmwiLCJhIjoiY2pueDgwZGhxMDVkbjN3cWx5dGlhd3p1eiJ9.FR_XylCvZZJLdB3No6Xxnw',
          styles: 'mapbox://styles/hadrienl/cjoplcnu821de2rs2cf0em4rw',
          center: [2.317600, 48.866500],
          zoom: 12.0,
          interactions: [{
            id: 'place-city-label-major',
            interaction: 'displayDetails',
          }],
        }]}
      />
    </div>
  </VisualizerProvider>
));
