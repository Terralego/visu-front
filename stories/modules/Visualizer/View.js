import React from 'react';

import InteractiveMap from '../../../modules/Map/InteractiveMap';

const layersTree = [{
  group: 'Limites administratives',
  layers: [{
    label: 'Departements',
    initialState: {
      active: false,
    },
    layers: ['terralego-departements'],
  }, {
    label: 'EPCI',
    initialState: {
      active: false,
    },
    layers: ['terralego-epci'],
  }],
}, {
  group: 'Périmètre d\'urbanisme',
  layers: [{
    label: 'SCOT',
    initialState: {
      active: false,
    },
    layers: ['terralego-scot'],
  }],
}, {
  group: 'Bâtiments',
  layers: [{
    label: 'Établissements',
    initialState: {
      active: false,
    },
    layers: ['terralego-etablissements'],
  }],
}, {
  group: 'EAE',
  layers: [{
    label: 'EAE',
    initialState: {
      active: false,
    },
    sublayers: [{
      label: 'Toutes',
      layers: ['terralego-eae'],
    }, {
      label: 'par vocation',
      layers: ['terralego-eae-sync'],
      legend: {
        items: [
          { label: 'Mixte', color: '#fe0200' },
          { label: 'Tertiaire supérieur', color: '#6fab46' },
          { label: 'Commerce de gros/Logistique', color: '#fec000' },
          { label: 'Construction', color: '#a4a6a4' },
          { label: 'Activités supports', color: '#8ea8db' },
          { label: 'Industrie', color: '#245e91' },
          { label: 'Commerce de détail', color: '#ec7c31' },
          { label: 'Services aux particuliers', color: '#9e470e' },
          { label: 'Autres', color: '#6a89cc' },
        ],
      },
    }],
  }],
}];

export default () => (
  <div style={{ height: '100vh' }}>
    <InteractiveMap
      type="map"
      layersTree={layersTree}
      accessToken="pk.eyJ1IjoibWFraW5hY29ycHVzIiwiYSI6ImNqY3E4ZTNwcTFta3ozMm80d2xzY29wM2MifQ.Nwl_FHrWAIQ46s_lY0KNiQ"
      backgroundStyle={[
        { label: 'Thème blanc', url: 'mapbox://styles/mapbox/light-v9' },
        { label: 'Thème noir', url: 'mapbox://styles/mapbox/dark-v9' },
        { label: 'satellite', url: 'mapbox://styles/mapbox/satellite-v9' },
      ]}
      center={[5.386195159396806, 43.30072210972415]}
      zoom={15}
      maxZoom={16}
      minZoom={11}
      interactions={[{
        id: 'terralego-eae',
        interaction: 'displayDetails',
        template: `
[{{nom_ppal}}](https://fiches.sud-foncier-eco.fr/espaces-d-activites/{{id_eae}})
* {{bbox}}
* {{comdeta_et}}
* {{comdetail_eff}}
* {{comgr_et}}
* {{comgros_eff}}
* {{const_eff}}
* {{const_et}}
* {{date_crea}}
* {{date_maj}}
* {{geom}}
* {{geom3857}}
* {{id}}
* {{id_eae}}
* {{id_scot}}
* {{id_zone}}
* {{identifier}}
* {{indus_eff}}
* {{indus_et}}
* {{insee_comm}}
* {{insee_epci}}
* {{layer_id}}
* {{logis_eff}}
* {{logis_et}}
* {{nb_emplois}}
* {{nom_ppal}}
* {{nonpres_eff}}
* {{nonpres_et}}
* {{parti_eff}}
* {{parti_et}}
* {{pres_eff}}
* {{pres_et}}
* {{ray_eae}}
* {{supp_eff}}
* {{supp_et}}
* {{surf_total}}
* {{tertisup_eff}}
* {{tertisup_et}}
* {{type_esp}}
* {{voc_decl}}
* {{voc_dom}}
* {{voc_synth}}
            `,
      }, {
        id: 'terralego-eae',
        interaction: 'displayTooltip',
        trigger: 'mouseover',
        template: `
# {{nom_ppal}}

* Surface : {{surf_total}}km2
* CP : {{insee_comm}}
* Commune : à trouver
`,
      }, {
        id: 'terralego-etablissements',
        interaction: 'displayTooltip',
        trigger: 'mouseover',
        template: 'Etablissement',
      }, {
        id: 'terralego-etablissements',
        interaction: 'displayDetails',
        template: 'Etablissement',
      }]}
      customStyle={{
        sources: [{
          id: 'terralego',
          type: 'vector',
          url: 'http://dev-tiles-paca.makina-corpus.net/api/layer/__nogroup__/tilejson',
        }],
        layers: [
          {
            id: 'cadastre',
            type: 'raster',
            minzoom: 14,
            source: {
              type: 'raster',
              tiles: [
                'https://gpp3-wxs.ign.fr/6ldzy836lrz6uea2qmzawkwe/geoportail/wmts?LAYER=CADASTRALPARCELS.PARCELS&EXCEPTIONS=text/xml&FORMAT=image/png&SERVICE=WMTS&VERSION=1.0.0&REQUEST=GetTile&STYLE=bdparcellaire_o&TILEMATRIXSET=PM&TILEMATRIX={z}&TILEROW={y}&TILECOL={x}',
              ],
              tileSize: 256,
            },
            layout: {
              visibility: 'none',
            },
          }, {
            type: 'line',
            source: 'terralego',
            id: 'terralego-departements',
            paint: {
              'line-color': '#66c2a4',
              'line-width': 2,
            },
            layout: {
              visibility: 'visible',
            },
            'source-layer': 'departements',
          }, {
            type: 'line',
            source: 'terralego',
            id: 'terralego-communes',
            paint: {
              'line-color': '#66c2a4',
              'line-opacity': 0.4,
              'line-width': 1,
            },
            layout: {
              visibility: 'visible',
            },
            'source-layer': 'communes',
          }, {
            type: 'line',
            source: 'terralego',
            id: 'terralego-scot',
            paint: {
              'line-color': '#fe9929',
              'line-width': 1,
            },
            layout: {
              visibility: 'none',
            },
            'source-layer': 'scot',
          }, {
            type: 'line',
            source: 'terralego',
            id: 'terralego-epci',
            paint: {
              'line-color': '#f768a1',
              'line-width': 2,
            },
            layout: {
              visibility: 'none',
            },
            'source-layer': 'epci',
          },
          {
            type: 'fill',
            source: 'terralego',
            id: 'terralego-eae',
            layout: {
              visibility: 'visible',
            },
            paint: {
              'fill-color': '#41b6c4',
              'fill-opacity': 0.4,
            },
            'source-layer': 'zae',
          },
          {
            type: 'fill',
            source: 'terralego',
            id: 'terralego-eae-sync',
            layout: {
              visibility: 'none',
            },
            paint: {
              'fill-color': [
                'match',
                ['get', 'voc_synth'],
                'Mixte', '#fe0200',
                'Tertiaire supérieur', '#6fab46',
                'Commerce de gros/Logistique', '#fec000',
                'Construction', '#a4a6a4',
                'Activités supports', '#8ea8db',
                'Industrie', '#245e91',
                'Commerce de détail', '#ec7c31',
                'Services aux particuliers', '#9e470e',
                '#6a89cc',
              ],
              'fill-opacity': 0.8,
            },
            'source-layer': 'zae',
          },
          {
            type: 'circle',
            source: 'terralego',
            id: 'terralego-eae-employment',
            layout: {
              visibility: 'none',
            },
            paint: {
              'circle-radius': [
                'interpolate', ['linear'], ['zoom'],
                8,
                ['case', ['has', 'nb_emplois'],
                  [
                    'case',
                    ['<', ['get', 'nb_emplois'], 600],
                    3,
                    ['<', ['get', 'nb_emplois'], 2300],
                    6,
                    ['<', ['get', 'nb_emplois'], 5200],
                    9,
                    ['<', ['get', 'nb_emplois'], 12000],
                    12,
                    15,
                  ],
                  0],
                16,
                ['case', ['has', 'nb_emplois'],
                  [
                    'case',
                    ['<', ['get', 'nb_emplois'], 600],
                    30,
                    ['<', ['get', 'nb_emplois'], 2300],
                    60,
                    ['<', ['get', 'nb_emplois'], 5200],
                    90,
                    ['<', ['get', 'nb_emplois'], 12000],
                    120,
                    150,
                  ],
                  0],
              ],
              'circle-color': [
                'case',
                ['has', 'nb_emplois'],
                [
                  'case',
                  ['<', ['get', 'nb_emplois'], 600],
                  '#ffffb2',
                  ['<', ['get', 'nb_emplois'], 2300],
                  '#fecc5c',
                  ['<', ['get', 'nb_emplois'], 5200],
                  '#fd8c3c',
                  ['<', ['get', 'nb_emplois'], 12000],
                  '#f03b1f',
                  '#bd0226',
                ],
                '#fff',
              ],
            },
            'source-layer': 'zae-centroid',
          }, {
            type: 'fill',
            source: 'terralego',
            id: 'terralego-industrie',
            paint: {
              'fill-color': '#ae8964',
              'fill-opacity': 0.4,
            },
            'source-layer': 'territoire_industrie',
          }, {
            type: 'circle',
            source: 'terralego',
            id: 'terralego-etablissements',
            'source-layer': 'etablissements',
          },
        ],
      }}
    />
  </div>
);
