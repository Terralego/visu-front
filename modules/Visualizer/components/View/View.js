import React from 'react';
import PropTypes from 'prop-types';

import WidgetMap from '../../widgets/Map';
import Details from '../Details';

import './styles.scss';

export class View extends React.Component {
  static propTypes = {
    widgets: PropTypes.arrayOf(PropTypes.shape({
      type: PropTypes.oneOf(['map']),
    })).isRequired,
    DetailsComponent: PropTypes.func,
  };

  static defaultProps = {
    DetailsComponent: Details,
  }

  state = {
    widgetsComponents: [],
  };

  componentDidMount () {
    this.generateWidgets();
  }

  componentDidUpdate ({ widgets: prevWidgets }) {
    const { widgets } = this.props;
    if (widgets !== prevWidgets) {
      this.generateWidgets();
    }
  }

  closeDetails = () => {
    const { setDetails } = this.props;
    setDetails(null);
  }

  generateWidgets () {
    const { widgets } = this.props;
    const widgetsComponents = widgets.map(({ type, ...props }, key) => {
      const widgetComponent = { ...props, key };
      switch (type) {
        case 'map':
          widgetComponent.Component = WidgetMap;
          break;
        default:
          throw new Error('Invalid widget type');
      }
      return widgetComponent;
    });
    this.setState({ widgetsComponents });
  }

  render () {
    const { DetailsComponent, details } = this.props;
    const { widgetsComponents } = this.state;
    const { closeDetails } = this;
    const visible = !!details;
    const { features: [{ properties }] = [{}], template = '' } = details || {};

    return (
      <div className="visualizer-view">
        {widgetsComponents.map(({ Component, key, ...props }) => (
          <Component
            key={key}
            {...props}
          />
        ))}
        <DetailsComponent
          visible={visible}
          template={template}
          onClose={closeDetails}
          {...properties}
        />
      </div>
    );
  }
}

export default View;
