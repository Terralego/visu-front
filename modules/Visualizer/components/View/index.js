import View from './View';
import { connectVisualizerProvider } from '../../services/context';

export default connectVisualizerProvider('details', 'setDetails')(View);
