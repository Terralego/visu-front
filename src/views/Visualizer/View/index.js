import React from 'react';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectView } from './context';
import View from './View';
import ViewProvider from './Provider';

const ConnectedView = connectAuthProvider('authenticated')(connectView(
  'initLayersState',
  'setLayersTreeState',
  'layersTreeState',
  'query',
  'map',
  'setMap',
  'setMapState',
  'mapIsResizing',
  'searchQuery',
  'setVisibleBoundingBox',
  'visibleBoundingBox',
)(View));

export default props => (
  <ViewProvider {...props}>
    <ConnectedView {...props} />
  </ViewProvider>
);
