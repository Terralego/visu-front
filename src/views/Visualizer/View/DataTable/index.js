import { connectLayersTree } from '@terralego/core/modules/Visualizer/LayersTree';

import DataTable from './DataTable';

export default connectLayersTree(({ layersTreeState, query, map, visibleBoundingBox }) => ({
  query,
  map,
  visibleBoundingBox,
  displayedLayer: Array
    .from(layersTreeState)
    .filter(([, { table }]) => table)
    .map(([layer, state]) => ({ ...layer, state }))[0],
}))(DataTable);
