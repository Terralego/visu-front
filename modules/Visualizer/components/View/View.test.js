import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { View } from './View';

import WidgetMap from '../../widgets/Map';

jest.mock('../../widgets/Map', () => jest.fn(() => null));

it('should render correctly', () => {
  const tree = renderer.create((
    <View
      widgets={[{ type: 'map' }]}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should build widgets', () => {
  const widgets = [{ type: 'map', foo: 'bar' }];
  const wrapper = shallow(<View
    widgets={widgets}
  />);

  const { widgetsComponents } = wrapper.state();
  expect(widgetsComponents.length).toBe(1);
  expect(widgetsComponents[0].Component).toBe(WidgetMap);
  expect(widgetsComponents[0].foo).toBe('bar');
  expect(widgetsComponents[0].key).toBe(0);
});

it('should regenerate widgets components', () => {
  const widgets = [{ type: 'map', foo: 'bar' }];
  const wrapper = shallow(<View
    widgets={widgets}
  />);
  const instance = wrapper.instance();
  instance.generateWidgets = jest.fn();
  wrapper.setProps({
    widgets: [{ type: 'map' }],
  });
  expect(instance.generateWidgets).toHaveBeenCalled();
});

it('should close details', () => {
  const setDetails = jest.fn();
  const instance = new View({ setDetails }, {});
  instance.closeDetails();
  expect(setDetails).toHaveBeenCalledWith(null);
});

it('should throw with invalid widget name', () => {
  class ViewWithoutProptypes {
    constructor (props) {
      this.props = props;
      this.generateWidgets = View.prototype.generateWidgets.bind(this);
    }
  }
  const instance = new ViewWithoutProptypes({ widgets: [{ type: 'foo' }] }, {});
  expect(() => instance.generateWidgets()).toThrow('Invalid widget type');
});
