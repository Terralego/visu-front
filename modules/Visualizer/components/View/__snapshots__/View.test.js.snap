// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="visualizer-view"
>
  <div
    className="view-details "
  >
    <button
      className="view-details__close-button"
      onClick={[Function]}
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-cross"
        icon="cross"
      >
        <svg
          data-icon="cross"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            cross
          </desc>
          <path
            d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 0 0-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 0 0-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 0 0 1.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 0 0 .71-1.71L9.41 8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
